mono/metadata/threads.c

```diff
@@ -4678,15 +4660,24 @@ static void CALLBACK dummy_apc (ULONG_PTR param)
  * Performs the operation that the requested thread state requires (abort,
  * suspend or stop)
  */
-static MonoExceptionHandle
-mono_thread_execute_interruption (void)
+static gboolean
+mono_thread_execute_interruption (MonoExceptionHandle *pexc)
 {
-	MONO_REQ_GC_UNSAFE_MODE;
+	// Use caller's frame if provided, else create one.
+	// This is not just an optimization -- it is needed to return
+	// a handle to the caller through an out-parameter.
 
-	HANDLE_FUNCTION_ENTER ();
+	if (!pexc) {
+		HANDLE_FUNCTION_ENTER ();
+		MonoExceptionHandle exc;
+		HANDLE_FUNCTION_RETURN_VAL (mono_thread_execute_interruption (&exc));
```

Comment:
Okay, so this `if (!pexc) {...}` block fixes callers who do `mono_thread_execute_interruption (NULL)` by using a local `MonoExceptionHandle` which will be assigned a garbage value by the 'real' `mono_thread_execute_interruption` call.

Reply:
This was the problem. VAL clears the frame and then evaluates its parameter, so the handle is lost.